File: Algorithms/0049_Group_Anagrams/Python/Group_Anagrams_Solution_1.py

# Space: O(n)
# Time: O(n)

import collections


class Solution:
    def groupAnagrams(self, strs):
        length = len(strs)
        if length == 1: return [strs]
        res = []
        cache = collections.defaultdict(list)

        for word in strs:
            key = ''.join(sorted([c for c in word]))
            cache[key].append(word)

        for key in cache:
            res.append(cache[key])
        return res

File: Algorithms/0049_Group_Anagrams/Java/GroupAnagramsSolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        if (strs.length == 1){
            List res = new ArrayList<>();
            res.add(Arrays.asList(strs));
        }
        List<List<String>> res = new ArrayList<>();
        Map<String, List<String>> cache = new HashMap<>();
        for (String s : strs) {
            String key = generateKey(s);
            cache.computeIfAbsent(key, k -> new ArrayList<>()).add(s);
        }
        for (String key : cache.keySet()) {
            res.add(cache.get(key));
        }
        return res;
    }

    private String generateKey(String str) {
        char[] chars = str.toCharArray();
        Arrays.sort(chars);
        return new String(chars);
    }
}




